File: backend/app/models/notification.py
import enum
from sqlalchemy import Column, String, Text, Boolean, Enum, DateTime, ForeignKey
from sqlalchemy.sql import func

from app.database.base import Base


class NotificationType(str, enum.Enum):
    EVAL_COMPLETE = "eval_complete"
    LIKE = "like"
    COMMENT = "comment"


class Notification(Base):
    __tablename__ = "notifications"

    id = Column(String(36), primary_key=True)
    user_id = Column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)

    type = Column(Enum(NotificationType), nullable=False)
    title = Column(String(200), nullable=False)
    message = Column(Text, nullable=False)

    # Related entities
    related_evaluation_id = Column(String(36))
    related_post_id = Column(String(36))
    related_user_id = Column(String(36))

    is_read = Column(Boolean, default=False, index=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)

File: frontend/lib/models/notification.dart
class AppNotification {
  final String id;
  final String type;
  final String title;
  final String message;
  final String? relatedEvaluationId;
  final String? relatedPostId;
  final String? relatedUserId;
  final bool isRead;
  final DateTime createdAt;

  const AppNotification({
    required this.id,
    required this.type,
    required this.title,
    required this.message,
    this.relatedEvaluationId,
    this.relatedPostId,
    this.relatedUserId,
    required this.isRead,
    required this.createdAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) =>
      AppNotification(
        id: json['id'] as String,
        type: json['type'] as String,
        title: json['title'] as String,
        message: json['message'] as String,
        relatedEvaluationId: json['related_evaluation_id'] as String?,
        relatedPostId: json['related_post_id'] as String?,
        relatedUserId: json['related_user_id'] as String?,
        isRead: json['is_read'] as bool? ?? false,
        createdAt: DateTime.parse(json['created_at'] as String),
      );

  AppNotification copyWith({bool? isRead}) => AppNotification(
        id: id,
        type: type,
        title: title,
        message: message,
        relatedEvaluationId: relatedEvaluationId,
        relatedPostId: relatedPostId,
        relatedUserId: relatedUserId,
        isRead: isRead ?? this.isRead,
        createdAt: createdAt,
      );
}
